73_ArithmeticBench/main.cpp

```diff
@@ -0,0 +1,949 @@
+				m_spirv_isa_cache = m_device->createPipelineCache(spirv_isa_cache_data);
+			}
+		}
+		{
+			// TODO: rename `deleteDirectory` to just `delete`? and a `IFile::setSize()` ?
+			m_system->deleteDirectory(spirv_isa_cache_path);
+			ISystem::future_t<smart_refctd_ptr<IFile>> fileCreate;
+			m_system->createFile(fileCreate,spirv_isa_cache_path,IFile::ECF_WRITE);
+			// I can be relatively sure I'll succeed to acquire the future, the pointer to created file might be null though.
+			m_spirv_isa_cache_output=*fileCreate.acquire();
+			if (!m_spirv_isa_cache_output)
+				logFail("Failed to Create SPIR-V to ISA cache file.");
+		}
+
+		// load shader source from file
+		auto getShaderSource = [&](const char* filePath) -> auto
+		{
+			IAssetLoader::SAssetLoadParams lparams = {};
+			lparams.logger = m_logger.get();
+			lparams.workingDirectory = "";
+			auto bundle = m_assetMgr->getAsset(filePath, lparams);
+			if (bundle.getContents().empty() || bundle.getAssetType()!=IAsset::ET_SHADER)
+			{
+				m_logger->log("Shader %s not found!", ILogger::ELL_ERROR, filePath);
+				exit(-1);
+			}
+			auto firstAssetInBundle = bundle.getContents()[0];
+			return smart_refctd_ptr_static_cast<ICPUShader>(firstAssetInBundle);
+		};
+
+		auto subgroupTestSource = getShaderSource("app_resources/testSubgroup.comp.hlsl");
+		auto subgroupBenchSource = getShaderSource("app_resources/benchmarkSubgroup.comp.hlsl");
+		//auto workgroupTestSource = getShaderSource("app_resources/testWorkgroup.comp.hlsl");
+		// now create or retrieve final resources to run our tests
+		sema = m_device->createSemaphore(timelineValue);
+		resultsBuffer = ICPUBuffer::create({ outputBuffers[0]->getSize() });
+		smart_refctd_ptr<IGPUCommandBuffer> cmdbuf;
+		{
+			smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(computeQueue->getFamilyIndex(),IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+			if (!cmdpool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY,{&cmdbuf,1}))
+			{
+				logFail("Failed to create Command Buffers!\n");
+				return false;
+			}
+		}
+		
+		// TODO variable items per invocation?
+		const uint32_t NumLoops = 1000u;
+		const std::array<uint32_t, 3> workgroupSizes = { 256, 512, 1024 };
+		// const auto MaxWorkgroupSize = m_physicalDevice->getLimits().maxComputeWorkGroupInvocations;
+		const auto MinSubgroupSize = m_physicalDevice->getLimits().minSubgroupSize;
+		const auto MaxSubgroupSize = m_physicalDevice->getLimits().maxSubgroupSize;
+		
+		if (b_runTests)
+		{
+			runTests(cmdbuf.get(), subgroupTestSource, elementCount, ItemsPerInvocation, MinSubgroupSize, MaxSubgroupSize, workgroupSizes);
+
+			m_logger->log("==========Result==========", ILogger::ELL_INFO);
+			m_logger->log("Fail Count: %u", ILogger::ELL_INFO, totalFailCount);
+		}
+
+		// for each workgroup size (manually adjust items per invoc, operation else uses up a lot of ram)
```

Comment:
if you only benchmark 1 op (plus) instead of all 7 or 8, you'll only have one output buffer.

This means you can do varying items per invocation tests as well

Reply:
bump